### ISSUE TITLE
[BUG] The object has been collected to prevent unbounded heap growth

### ISSUE THREAD

**Issue by tiangolo** [microsoft/playwright#28503]
### System info
- Playwright Version: v1.40.0
- Operating System: macOS 13.5.2
- Browser: Chromium, Firefox
- Other info:
	- Python: 3.12

### Source code

```Python
from playwright.sync_api import sync_playwright


url = "https://testdriven.io/opensource/"


with sync_playwright() as p:
    browser = p.firefox.launch(headless=False)
    page = browser.new_page()
    page.goto(url)
    table = page.query_selector("table.table.dataTable")
    assert table
    headers = table.query_selector_all("thead tr th")
    headers_text = [header.inner_text() for header in headers]
    assert headers_text[1:] == ["Project", "Amount", "Date", "Notes"]
    rows = table.query_selector_all("tbody tr")
    rows_data: list[list[str]] = []
    for row in rows:
        data: list[str] = []
        tds = row.query_selector_all("td")
        for cell in tds:
            data.append(cell.inner_text())
        rows_data.append(data)
print(rows_data)
```

**Steps**

- Install playwright (probably in a venv):

```console
$ pip install playwright
$ playwright install
```

- Put that script in a file `main.py`
- Run the script:

```console
$ python main.py

Traceback (most recent call last):
  File "/Users/user/code/demo/main.py", line 20, in <module>
    tds = row.query_selector_all("td")
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/sync_api/_generated.py", line 2917, in query_selector_all
    self._sync(self._impl_obj.query_selector_all(selector=selector))
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/_impl/_sync_base.py", line 115, in _sync
    return task.result()
           ^^^^^^^^^^^^^
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/_impl/_element_handle.py", line 324, in query_selector_all
    await self._channel.send("querySelectorAll", dict(selector=selector)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/_impl/_connection.py", line 62, in send
    return await self._connection.wrap_api_call(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/_impl/_connection.py", line 492, in wrap_api_call
    return await cb()
           ^^^^^^^^^^
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/_impl/_connection.py", line 84, in inner_send
    callback = self._connection._send_message_to_server(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user/code/demo/.venv/lib/python3.12/site-packages/playwright/_impl/_connection.py", line 320, in _send_message_to_server
    raise Error(
playwright._impl._errors.Error: The object has been collected to prevent unbounded heap growth.
```

**Expected**

The intended purpose, and it was working until the past month, was to load that page: https://testdriven.io/opensource/, read the table, and collect the data in the table on that page.

Then it would just print a list of lists (my actual code actually processes the data further, but I'm putting here the minimum example to show the error).

**Actual**

It errors out saying that `The object has been collected to prevent unbounded heap growth`.

---

I don't really see something obviously wrong in my code that could lead to unbound memory, and I don't see an API that would release resources as I go or something similar, I'm not really sure how else to debug it.

**Comment by Jamesernator** [comment]
I've been having the same problem, with basically the same setup (though in JS): a loop over a list of handles returned by `elementHandle.$$` (i.e. query_selector_all). Memory usage isn't high at all, it just happens after a few iterations of the loop fairly consistently.

**Comment by Jamesernator** [comment]
Looking into the culprit, the cause is the addition of this highly arbitrary limit in [dispatcher.ts](https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/server/dispatchers/dispatcher.ts#L37):

```ts
let maxDispatchers = 1000;
export function setMaxDispatchersForTest(value: number | undefined) {
  maxDispatchers = value || 1000;
}
```

This makes it functionally useless if a `.$$`/`.query_selector_all` happens to return a list of handles more than 1000 (considerably less if you create additional handles in the loop).

**Comment by mxschmitt** [comment]
.$, $$, aka. query_selector/query_selector_all is actually discouraged because of that, because we from the Playwright perspective can never tell if the object is still in use or not without relying on concepts like finalisers. You can manually call `dispose()` on it to dispose on it. Other reasons are that `ElementHandlers` represent a single DOM node which would lead to side effects if the app is hydrating e.g.

TL:DR: We strongly discourage `ElementHandles` and recommend `Locators` instead as per [here](https://playwright.dev/python/docs/api/class-page#page-query-selector-all). Locators don't belong to a single DOM node and instead resolve the element on the action (click, fill etc.) as per [here](https://playwright.dev/python/docs/api/class-locator).

Let us know if that works for you.

**Comment by Jamesernator** [comment]
> because we from the Playwright perspective can never tell if the object is still in use or not without relying on concepts like finalisers

Why does Playwright not just use finalizers? They are supported in all of Playwright's targeted languages ([JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry), [Python](https://docs.python.org/3/library/weakref.html#weakref.finalize), [Java](https://www.oracle.com/technical-resources/articles/javase/finalization.html), [C#](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/finalizers)).

> TL:DR: We strongly discourage `ElementHandles` and recommend `Locators` instead as per [here](https://playwright.dev/python/docs/api/class-page#page-query-selector-all). Locators don't belong to a single DOM node and instead resolve the element on the action (click, fill etc.) as per [here](https://playwright.dev/python/docs/api/class-locator).
> 
> Let us know if that works for you.

I updated Playwright mainly for running on newer browser versions, I did not expect such an update to actively break existing code. I don't really feel like going and updating everything to use a different API when the current code works perfectly fine.

For now, I think I'll use just [used `patch-package`](https://www.npmjs.com/package/patch-package) to delete the arbitrary limitation added.

**Comment by dgozman** [comment]
Fixed by #28629.
